### ISSUE TITLE
qt-main has missing dependencies that can cause segmentation faults on linux

### ISSUE THREAD

**Issue by erykoff** [conda-forge/qt-main-feedstock#32]
### Solution to issue cannot be found in the documentation.

- [X] I checked the documentation.

### Issue

If you create a new env with pyqt and qtmain (5.15) and matplotlib then you _may_ have a segmentation fault when trying to plot with the default `QtAgg` backend.  For example create a new env on linux x86 with  `mamba create -n qttest python=3.10 pyqt qt-main matplotlib`.

Then run the following in the `qttest` env:
```
import matplotlib
import matplotlib.pyplot as plt

print(matplotlib.get_backend())

plt.plot(1, 2)
plt.show()
```

After seeing that `QtAgg` is the default backend you _may_ get an `Abort` when loading the xcb plugin with the dreaded `qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.` error.

I tracked this down to the fact that somewhere in between `qt=5.12` and `qt-main=5.15.3` there are more dependencies for the xcb plugin.  Running `ldd $CONDA_PREFIX/plugins/platforms/libqxcb.so` shows that there are a number of linked libraries that may not be installed.  These _may_ be installed on your system (via yum on centos7 for example) which causes things to run fine, or if they're not installed ... 💥 .

On a system _with_ these libraries installed, I see that:

```
libX11-xcb.so.1 => /lib64/libX11-xcb.so.1 (0x00007f70fc98d000)
libxcb-icccm.so.4 => /lib64/libxcb-icccm.so.4 (0x00007f70fc788000)
libxcb-image.so.0 => /lib64/libxcb-image.so.0 (0x00007f70fc583000)
libxcb-keysyms.so.1 => /lib64/libxcb-keysyms.so.1 (0x00007f70fc380000)
libxcb-render-util.so.0 => /lib64/libxcb-render-util.so.0 (0x00007f70fc169000)
```

(And on a system without these libraries installed these say `not found`.

`icccm` is provided by `xcb-util-wm`
`keysyms` is provided by `xcb-util-keysyms`
`image` is provided by `xcb-util-image`
`render-util` is provided by `xcb-util-renderutil`

The `xcb-util-wm` and `xcb-util-keysyms` packages appear to be available on conda-forge and should be dependencies of `qt-main`.  However, `xcb-image` and `xcb-util-renderutil` do not seem to be available on conda forge, as far as I can tell.

It may also be that these should not be dependencies at all, and there is a problem with the configure script settings in `qt-main` that is turning on options that are not intentional.  I have not investigated this.

In any event, as more envs are updated this is going to bite more people who will be very confused.

### Installed packages

```shell
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       2_gnu    conda-forge
alsa-lib                  1.2.3.2              h166bdaf_0    conda-forge
attr                      2.5.1                h166bdaf_0    conda-forge
brotli                    1.0.9                h166bdaf_7    conda-forge
brotli-bin                1.0.9                h166bdaf_7    conda-forge
bzip2                     1.0.8                h7f98852_4    conda-forge
ca-certificates           2022.5.18.1          ha878542_0    conda-forge
certifi                   2022.5.18.1     py310hff52083_0    conda-forge
cycler                    0.11.0             pyhd8ed1ab_0    conda-forge
dbus                      1.13.6               h5008d03_3    conda-forge
expat                     2.4.8                h27087fc_0    conda-forge
fftw                      3.3.10          nompi_h77c792f_102    conda-forge
font-ttf-dejavu-sans-mono 2.37                 hab24e00_0    conda-forge
font-ttf-inconsolata      3.000                h77eed37_0    conda-forge
font-ttf-source-code-pro  2.038                h77eed37_0    conda-forge
font-ttf-ubuntu           0.83                 hab24e00_0    conda-forge
fontconfig                2.14.0               h8e229c2_0    conda-forge
fonts-conda-ecosystem     1                             0    conda-forge
fonts-conda-forge         1                             0    conda-forge
fonttools                 4.33.3          py310h5764c6d_0    conda-forge
freetype                  2.10.4               h0708190_1    conda-forge
gettext                   0.19.8.1          h73d1719_1008    conda-forge
giflib                    5.2.1                h36c2ea0_2    conda-forge
gst-plugins-base          1.20.2               hcf0ee16_0    conda-forge
gstreamer                 1.20.2               hd4edc92_1    conda-forge
icu                       70.1                 h27087fc_0    conda-forge
jack                      1.9.18            hfd4fe87_1001    conda-forge
jpeg                      9e                   h166bdaf_1    conda-forge
keyutils                  1.6.1                h166bdaf_0    conda-forge
kiwisolver                1.4.2           py310hbf28c38_1    conda-forge
krb5                      1.19.3               h3790be6_0    conda-forge
lcms2                     2.12                 hddcbb42_0    conda-forge
ld_impl_linux-64          2.36.1               hea4e1c9_2    conda-forge
lerc                      3.0                  h9c3ff4c_0    conda-forge
libblas                   3.9.0           14_linux64_openblas    conda-forge
libbrotlicommon           1.0.9                h166bdaf_7    conda-forge
libbrotlidec              1.0.9                h166bdaf_7    conda-forge
libbrotlienc              1.0.9                h166bdaf_7    conda-forge
libcap                    2.64                 ha37c62d_0    conda-forge
libcblas                  3.9.0           14_linux64_openblas    conda-forge
libclang                  14.0.4          default_h2e3cab8_0    conda-forge
libclang13                14.0.4          default_h3a83d3e_0    conda-forge
libcups                   2.3.3                hf5a7f15_1    conda-forge
libdb                     6.2.32               h9c3ff4c_0    conda-forge
libdeflate                1.10                 h7f98852_0    conda-forge
libedit                   3.1.20191231         he28a2e2_2    conda-forge
libevent                  2.1.10               h9b69904_4    conda-forge
libffi                    3.4.2                h7f98852_5    conda-forge
libflac                   1.3.4                h27087fc_0    conda-forge
libgcc-ng                 12.1.0              h8d9b700_16    conda-forge
libgfortran-ng            12.1.0              h69a702a_16    conda-forge
libgfortran5              12.1.0              hdcd56e2_16    conda-forge
libglib                   2.70.2               h174f98d_4    conda-forge
libgomp                   12.1.0              h8d9b700_16    conda-forge
libiconv                  1.16                 h516909a_0    conda-forge
liblapack                 3.9.0           14_linux64_openblas    conda-forge
libllvm14                 14.0.4               he0ac6c6_0    conda-forge
libnsl                    2.0.0                h7f98852_0    conda-forge
libogg                    1.3.4                h7f98852_1    conda-forge
libopenblas               0.3.20          pthreads_h78a6416_0    conda-forge
libopus                   1.3.1                h7f98852_1    conda-forge
libpng                    1.6.37               h21135ba_2    conda-forge
libpq                     14.3                 hd77ab85_0    conda-forge
libsndfile                1.0.31               h9c3ff4c_1    conda-forge
libstdcxx-ng              12.1.0              ha89aaad_16    conda-forge
libtiff                   4.4.0                h0fcbabc_0    conda-forge
libtool                   2.4.6             h9c3ff4c_1008    conda-forge
libudev1                  249                  h166bdaf_2    conda-forge
libuuid                   2.32.1            h7f98852_1000    conda-forge
libvorbis                 1.3.7                h9c3ff4c_0    conda-forge
libwebp                   1.2.2                h3452ae3_0    conda-forge
libwebp-base              1.2.2                h7f98852_1    conda-forge
libxcb                    1.13              h7f98852_1004    conda-forge
libxkbcommon              1.0.3                he3ba5ed_0    conda-forge
libxml2                   2.9.14               h22db469_0    conda-forge
libzlib                   1.2.12               h166bdaf_0    conda-forge
lz4-c                     1.9.3                h9c3ff4c_1    conda-forge
matplotlib                3.5.2           py310hff52083_0    conda-forge
matplotlib-base           3.5.2           py310h5701ce4_0    conda-forge
munkres                   1.1.4              pyh9f0ad1d_0    conda-forge
mysql-common              8.0.29               haf5c9bc_1    conda-forge
mysql-libs                8.0.29               h28c427c_1    conda-forge
ncurses                   6.3                  h27087fc_1    conda-forge
nspr                      4.32                 h9c3ff4c_1    conda-forge
nss                       3.78                 h2350873_0    conda-forge
numpy                     1.22.4          py310h4ef5377_0    conda-forge
openjpeg                  2.4.0                hb52868f_1    conda-forge
openssl                   1.1.1o               h166bdaf_0    conda-forge
packaging                 21.3               pyhd8ed1ab_0    conda-forge
pcre                      8.45                 h9c3ff4c_0    conda-forge
pillow                    9.1.1           py310he619898_1    conda-forge
pip                       22.1.1             pyhd8ed1ab_0    conda-forge
pthread-stubs             0.4               h36c2ea0_1001    conda-forge
pulseaudio                14.0                 hbc9ff1d_7    conda-forge
pyparsing                 3.0.9              pyhd8ed1ab_0    conda-forge
pyqt                      5.15.4          py310hd8f1fbe_0    conda-forge
pyqt5-sip                 12.9.0          py310hd8f1fbe_0    conda-forge
python                    3.10.4          h9a8a25e_0_cpython    conda-forge
python-dateutil           2.8.2              pyhd8ed1ab_0    conda-forge
python_abi                3.10                    2_cp310    conda-forge
qt-main                   5.15.3               hf97cb25_1    conda-forge
readline                  8.1                  h46c0cb4_0    conda-forge
setuptools                62.3.2          py310hff52083_0    conda-forge
sip                       6.5.1           py310h122e73d_2    conda-forge
six                       1.16.0             pyh6c4a22f_0    conda-forge
sqlite                    3.38.5               h4ff8645_0    conda-forge
tk                        8.6.12               h27826a3_0    conda-forge
toml                      0.10.2             pyhd8ed1ab_0    conda-forge
tornado                   6.1             py310h5764c6d_3    conda-forge
tzdata                    2022a                h191b570_0    conda-forge
unicodedata2              14.0.0          py310h5764c6d_1    conda-forge
wheel                     0.37.1             pyhd8ed1ab_0    conda-forge
xorg-libxau               1.0.9                h7f98852_0    conda-forge
xorg-libxdmcp             1.1.3                h7f98852_0    conda-forge
xz                        5.2.5                h516909a_1    conda-forge
zlib                      1.2.12               h166bdaf_0    conda-forge
zstd                      1.5.2                h8a70e8d_1    conda-forge
```
```


### Environment info

```shell
active environment : qttest
    active env location : /project/erykoff/test_conda/miniforge3/envs/qttest
            shell level : 3
       user config file : /home/erykoff/.condarc
 populated config files : /project/erykoff/test_conda/miniforge3/.condarc
          conda version : 4.12.0
    conda-build version : not installed
         python version : 3.9.10.final.0
       virtual packages : __linux=3.10.0=0
                          __glibc=2.17=0
                          __unix=0=0
                          __archspec=1=x86_64
       base environment : /project/erykoff/test_conda/miniforge3  (writable)
      conda av data dir : /project/erykoff/test_conda/miniforge3/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /project/erykoff/test_conda/miniforge3/pkgs
                          /home/erykoff/.conda/pkgs
       envs directories : /project/erykoff/test_conda/miniforge3/envs
                          /home/erykoff/.conda/envs
               platform : linux-64
             user-agent : conda/4.12.0 requests/2.27.1 CPython/3.9.10 Linux/3.10.0-1160.49.1.el7.x86_64 centos/7.9.2009 glibc/2.17
                UID:GID : 52876:202
             netrc file : None
           offline mode : False
```
```

**Comment by isuruf** [comment]
These are added as CDT packages which means the user has to have these in their system.

**Comment by erykoff** [comment]
Where is the documentation on that?

**Comment by erykoff** [comment]
I note that on this feedstock the {{devel}} versions of the packages are listed as cdt packages for the build, but nowhere are the non-devel versions listed (which are required to run and I assume also test).  I would like to make sure that these are listed clearly somewhere.  Thanks!

**Comment by ccordoba12** [comment]
> Where is the documentation on that?

https://conda-forge.org/docs/maintainer/knowledge_base.html#core-dependency-tree-packages-cdts

> I note that on this feedstock the {{devel}} versions of the packages are listed as cdt packages for the build, but nowhere are the non-devel versions listed

They are not required as part of the recipe. As @isuruf said, it's the user's responsibility to install them.

For testing with CentOS 7, they are listed here though:

https://github.com/conda-forge/qt-main-feedstock/blob/main/recipe/yum_requirements.txt

**Comment by erykoff** [comment]
Looking at that list of requirements in `yum_requirements.txt`, it has nothing to do with `xcb`.  In fact, `libxcb` is listed as a regular runtime dependency in meta.yaml for both `qt` and `qt-main`, because it is packaged with conda forge.  And until the latest version(s), that has been sufficient to use the `QtAgg` backend in matplotlib via pyqt.

My point is that now with `qt-main=5.13.5` there are additional packages that are required dependencies at the *runtime* (separate from build time), `xcb-util-wm` and `xcb-util-keysyms` (both currently available as regular packages on conda-forge) and `xcb-util-image` and `xcb-util-renderutil` (not available as regular packages on conda-forge but I don't think it would be difficult to add them, following the other `xcb` templates).  I still can't find where these dependencies are listed, even as CDT packages.  The only way I figured out which packages were missing was to have a system where they happened to be installed so I could do a `rpm -qif` to find out which libraries they came from.

**Comment by ccordoba12** [comment]
> Looking at that list of requirements in yum_requirements.txt, it has nothing to do with xcb

They are installed as transitive dependencies of the packages listed there. And that's easily verifiable because this feedstock has several Qt tests and they don't segfault.

> My point is that now with qt-main=5.13.5 there are additional packages that are required dependencies at the runtime

Sure, and as I said before the user has to install them.

> I still can't find where these dependencies are listed, even as CDT packages

Correct and (if I'm not mistaken) the purpose of that is to not create the Conda-forge Linux distro. In other words, to not have to create conda packages for every required Linux dependency.

> The only way I figured out which packages were missing was to have a system where they happened to be installed so I could do a rpm -qif to find out which libraries they came from.

Sure, that's good too.

**Comment by erykoff** [comment]
First of all, regarding the segfaults, I believe the fault here (that it segfaults rather than simply raises) is with either `pyqt` or `matplotlib`.

Second, asking every user who runs into this problem to spend many hours googling solutions before finding a thread to know what to turn on for QT debugging, and then finding that there are missing libraries with `ldd`, and then finding a working system to know where those libraries come from is not workable nor scalable.

Prior to April 22, 2022, everything needed to run qt came from conda forge (including `libxcb`).  Now, that is no longer the case, and this issue #32 may be the only place to find out what is necessary to run qt (which is installed when you install `jupyter` which is a fairly common operation).  If this is intentional, then I am requesting that it be documented somewhere.  (Or please point me to where it is documented if I have missed something!) Should I file an issue on https://conda-forge.org/docs/maintainer/knowledge_base.html (but I actually don't know where to file issues on that).

**Comment by hmaarrfk** [comment]
@erykoff thank you for hunting down the packages.  I have to say that I had to add the equivalent of

```
sudo yum install xcb-util-wm xcb-util-keysyms xcb-util-image xcb-util-renderutil
```
to a CI of mine today to get it to work.

To recreate, you can follow with a local linux machine:

<details>


```bash
docker run -it quay.io/condaforge/linux-anvil-cos7-x86_64


sudo yum install --assumeyes \
alsa-lib \
libXScrnSaver \
libXtst \
libXi \
libXcursor \
libXcomposite \
mesa-libGL \
mesa-dri-drivers \
libselinux \
libXdamage \
libXxf86vm \
libXext \
libX11-devel \
libXt-devel \
libXext-devel \
chrpath \
libXrender-devel \
gtk2-devel \
dbus-devel \
libSM-devel \
libICE-devel \
xorg-x11-server-Xvfb \
ruby 

# Create a virtual buffer
Xvfb :99 -screen 0 1920x1080x24 -dpi 96 -ac -nolisten tcp -nolisten unix &
export DISPLAY=:99

mamba install "pyside2>=5.15" --yes
python -c "from PySide2 import QtWidgets; app = QtWidgets.QApplication()"
# qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
# This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
#
# Available platform plugins are: eglfs, minimal, minimalegl, offscreen, vnc, webgl, xcb.
#
# Aborted (core dumped)

# install yum_requirements from pyside2
sudo yum install --assumeyes \
mesa-libGL-devel \
mesa-libGLU-devel \
mesa-libEGL-devel \
libX11-devel \
libXt-devel \
libXtst-devel \
libXrender-devel \
libXext-devel \
libXdmcp-devel \
libXcomposite-devel \
libXcursor-devel \
libxcb-devel \
libXau-devel \
libXi-devel \
libXScrnSaver-devel \
libXrandr-devel \
alsa-lib-devel \
xcb-util-devel \
xorg-x11-server-Xvfb \
pciutils-devel

python -c "from PySide2 import QtWidgets; app = QtWidgets.QApplication()"

# qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
# This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
# 
# Available platform plugins are: eglfs, minimal, minimalegl, offscreen, vnc, webgl, xcb.
# 
# Aborted (core dumped)

sudo yum install --assumeyes\
    xcb-util-wm xcb-util-keysyms xcb-util-image xcb-util-renderutil

python -c "from PySide2 import QtWidgets; app = QtWidgets.QApplication()"
# Success!
# QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-conda'
```
</details>

I don't really know what the best practice is for X11 dependencies. But anyway, I think @erykoff is correct in the sense that there is something missing from our CI tests or undocumented in `yum_requirement.txt`

**Comment by ryanvolz** [comment]
Just chiming in to say that I have been bitten by this in the `qt-main 5.15` migration, with failing CI tests that use Xvfb to run a pyqt program. For native linux builds, adding the `xcb-util*` packages to `yum_requirements.txt` works to get the test to pass. For cross-compiled linux builds though (i.e. linux-aarch64 on Azure), that doesn't work because it installs the x86_64 RPMs for the build image. Having the `xcb-util*` packages available as regular packages and not as CDTs would make it possible to get the needed libraries into the host environment.

I understand the purpose of CDTs and their use for Qt, but they assume that the CDT package is such an integral part of the desired functionality of the base system (in this case, X11) that everyone will have it already. That assumption seems to not hold for `xcb-util*` in the wild, or even in the conda-forge CI without special care. That says to me that it would be best to [add the 2 missing packages to conda-forge](https://github.com/conda-forge/staged-recipes/pull/19156) and add all 5 of the `xcb-util*` packages as non-CDT dependencies to `qt-main`. Moreover, `libxcb` is already a proper non-CDT dependency, so it seems backward that the required `xcb-util*` on top of it would *not* be.

**Comment by erykoff** [comment]
Following up on @ryanvolz 's helpful comment, at the very least the `yum_requirements.txt` is currently under-specified and this implies the test coverage is insufficient.  What I did was make a fresh Docker download of centos7, install all the yum_requirements, download miniforge, and create an env with qt-main.  In the end, the `libqxcb.so` file has a bunch of missing libraries.

Specifically:
```
sudo docker run -it centos:7 bash

xargs yum -y install < yum_requirements.txt

curl -OL https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-Linux-x86_64.sh

bash Miniforge3-Linux-x86_64.sh

source /root/miniforge3/etc/profile.d/conda.sh
conda activate base
conda create -n qtmain qt-main

conda activate qtmain

ldd $CONDA_PREFIX/plugins/platforms/libqxcb.so | grep "not found"
	libxcb-icccm.so.4 => not found
	libxcb-image.so.0 => not found
	libxcb-keysyms.so.1 => not found
	libxcb-render-util.so.0 => not found
	libxcb-icccm.so.4 => not found
	libxcb-image.so.0 => not found
	libxcb-keysyms.so.1 => not found
	libxcb-render-util.so.0 => not found
```

**Comment by isuruf** [comment]
> For cross-compiled linux builds though (i.e. linux-aarch64 on Azure), that doesn't work because it installs the x86_64 RPMs for the build image.

This is not true. We install them for the cross compiled platforms too. (Not using yum, but using the CDT packages specified in yum_requirements.txt). However as @erykoff said, the CDTs might be under-specificed or the dependencies of CDTs might be wrong.

**Comment by ryanvolz** [comment]
> This is not true. We install them for the cross compiled platforms too. (Not using yum, but using the CDT packages specified in yum_requirements.txt).

Ah, sorry, my mistake! I see where that happens now in the logs, and why it still wasn't working for me (missing `xcb-util` in `yum_requirements.txt`, which gets installed as a dependency by yum but not when using the CDT packages).

> However as @erykoff said, the CDTs might be under-specificed or the dependencies of CDTs might be wrong.

At the very least, the `xcb-util-image` RPM depends on `xcb-util`, but the same is not true for the CDTs. I think that is not the only place where the CDT dependencies don't match the RPMs within the X11 packages generally, but it is what got me here.

Overall though, I would agree that the current tests don't seem to exercise any of the parts of the libraries that require `xcb-util*`, or else they would be needed in `qt-main`'s `yum_requirements.txt`.

(I think there's a case to be made that all of the CDT and yum wrangling around X11 for packages that use Qt argues for using the available non-CDT packages, but that's a larger argument that would involve [renewing discussion](https://github.com/conda-forge/cfep/pull/7) of [CFEP 4](https://github.com/conda-forge/cfep/blob/main/cfep-04.md)).

**Comment by hoechenberger** [comment]
Just chiming in to let you know we've been bitten by this in our project / CI runs too. Took me a bit of googling to find this issue here.

**Comment by gillins** [comment]
Sorry if this is a silly suggestion, but what about creating an [activate script](https://docs.conda.io/projects/conda-build/en/latest/resources/activate-scripts.html) that checks these libs are available and prints a warning if not?

**Comment by erykoff** [comment]
Fixed with #39
